drivers: auxdisplay: Add TM1637 7-segment display driver

heronet opened the pull request:
Add a new auxdisplay driver for TM1637 7-segment LED displays. The driver supports:

- 4-digit 7-segment display output
- Decimal point positioning
- Brightness control (0-7 levels)
- Display on/off control
- Alphabet character support (A-Z, a-z) and digits (0-9)
- Basic cursor positioning

The driver implements the standard AUXDISPLAY API

kartben reviewed (dismissed):
you shouldn't be requiring a custom, device-specific, API for this - please implement this as an auxdisplay driver. Custom sample will likely also go away in the process

Thanks!

thedjnK commented:
Also there is a lot of changing of GPIO mode from input to output, would it be better to just configure it at start-up as output then to input and output again in that function, and optionally in on/off functions?

heronet commented on dts/bindings/auxdisplay/titanmicro,tm1637.yaml:
```diff
@@ -0,0 +1,34 @@
+  bit-delay-us:
+    type: int
+    default: 100
+    description: Bit delay in microseconds (default 100 µs)
```
the data sheet specifies minimum timings of 100ns - 400ns but I used 100us to account for GPIO behaviors across different platforms. I will elaborate more in the description

heronet commented:
> Also there is a lot of changing of GPIO mode from input to output, would it be better to just configure it at start-up as output then to input and output again in that function, and optionally in on/off functions?

Yes. That is what I should have done in the first place

simonguinot reviewed (dismissed):
Hello @heronet,

Thanks for this PR.

Please find my review below !

heronet commented:
Hi @thedjnK, @mbolivar, @kartben, @nashif, @xingrz, and other reviewers,

I wanted to check in on this PR for the TM1637 7-segment display driver. Please let me know if there are any remaining changes or feedback required to help move it forward. This is my first time sumbitting a driver so that would make sense.
Thank you for your time and reviews!

xingrz commented on drivers/auxdisplay/auxdisplay_tm1637.c:
```diff
@@ -0,0 +1,380 @@
+	0x5B, /* Z - same as 2 */
+};
+
+static const uint8_t minus_code = 0x40; /* Segment G only */
+#define DP_BIT (0x80)
```
```c
#define MINUS_BIT BIT(6)
#define DP_BIT    BIT(7)
```

Also, it would be clearer to add `#define BLANK (0)`.

xingrz commented on drivers/auxdisplay/auxdisplay_tm1637.c:
```diff
@@ -0,0 +1,380 @@
+			segment_code = segment_codes[c - '0'];
+			valid_char = true;
+		} else if (c >= 'A' && c <= 'Z') {
+			segment_code = segment_codes[c - 'A' + 10]; /* A starts at index 10 */
+			valid_char = true;
+		} else if (c >= 'a' && c <= 'z') {
+			segment_code = segment_codes[c - 'a' + 10]; /* a starts at index 10 */
+			valid_char = true;
```
Using `+ 10` for letters is not very clear. It’d be better to separate digit and alphabet segment codes into different arrays for readability.

xingrz commented on drivers/auxdisplay/auxdisplay_tm1637.c:
```diff
@@ -0,0 +1,380 @@
+			segment_code = segment_codes[c - 'a' + 10]; /* a starts at index 10 */
+			valid_char = true;
+		} else if (c == '-') {
+			segment_code = minus_code;
```
`MINUS_BIT` as mentioned above.

xingrz commented on drivers/auxdisplay/auxdisplay_tm1637.c:
```diff
@@ -0,0 +1,380 @@
+			segment_code = minus_code;
+			valid_char = true;
+		} else if (c == ' ') {
+			segment_code = 0;
```
`BLANK` as mentioned above.

xingrz commented on drivers/auxdisplay/auxdisplay_tm1637.c:
```diff
@@ -0,0 +1,380 @@
+
+/* Display control register bits */
+#define TM1637_DISPLAY_ON_BIT  BIT(3)
+#define TM1637_BRIGHTNESS_MASK 0x07
```
`BIT_MASK(3)`

xingrz reviewed (commented):
Thanks, looks good! Just a minor thought - it might be a bit clearer to define the segment codes using `BIT(x)` combinations instead of magic numbers. But it’s totally fine to keep it as is if you prefer.

heronet commented:
Hi @xingrz,

Thanks for taking a look again! I kept them in hex since it follows the standard convention for 7-segment lookup tables (e.g., https://en.wikichip.org/wiki/seven-segment_display/representing_letters). The hex format is more compact and matches how most 7-segment drivers represent these patterns.

Happy to change to BIT() macros if you prefer though.

Thanks

mbolivar commented on drivers/auxdisplay/Kconfig.tm1637:
```diff
@@ -0,0 +1,16 @@
+
+config AUXDISPLAY_TM1637
+	bool "Titan Micro TM1637 7-segment display support"
+	depends on DT_HAS_TITANMICRO_TM1637_ENABLED
```
Looks like a missing `default y` here

heronet committed:
dts: bindings: auxdisplay: Add TM1637 binding

Add device tree binding for TitanMicro TM1637 7-segment LED
display controller. The TM1637 uses a GPIO bit-banging protocol
with clock and data pins.

Properties:
- clk-gpios: Clock GPIO pin
- dio-gpios: Data I/O GPIO pin
- bit-delay-us: Bit delay for GPIO protocol timing

Signed-off-by: Siratul Islam <sirat4757@gmail.com>

heronet commented:
> > Hi @JarmouniA @thedjnK. I addressed the issues you mentioned. Would you please take another look when you get the time. Thanks.
> 
> [8ab82f9#r2392594184](https://github.com/zephyrproject-rtos/zephyr/commit/8ab82f9538b1782fb8bb891a0d33a164dab211bb#r2392594184) this comment is still outstanding

I'm not sure which comment the link is refering to(broken?) but I did some modifications and replied to one of your previous comments. Do you have an opinion on that? Thanks

thedjnK commented:
> > > Hi @JarmouniA @thedjnK. I addressed the issues you mentioned. Would you please take another look when you get the time. Thanks.
> > 
> > 
> > [8ab82f9#r2392594184](https://github.com/zephyrproject-rtos/zephyr/commit/8ab82f9538b1782fb8bb891a0d33a164dab211bb#r2392594184) this comment is still outstanding
> 
> I'm not sure which comment the link is refering to(broken?) but I did some modifications and replied to one of your previous comments. Do you have an opinion on that? Thanks

Let me try getting a link again https://github.com/zephyrproject-rtos/zephyr/pull/96510/commits/8ab82f9538b1782fb8bb891a0d33a164dab211bb#diff-e77c05676791af56edbb70e77ae08a9d2a29a04fb33642c40dc716f2198aa555R48 it's the characters, let's not add a hacky incomplete character set and stick to numerical only.
Hmm seems to be a github bug, if I open the link manually it goes to the line and shows the comments, if I click the same link but when I've posted on github, it just goes to the commit, without comments, highlights the line but does not scroll to it, basically click commits, click the second commit, scroll down to line 48 in the .c file

heronet commented:
> > > > Hi @JarmouniA @thedjnK. I addressed the issues you mentioned. Would you please take another look when you get the time. Thanks.
> > > 
> > > 
> > > [8ab82f9#r2392594184](https://github.com/zephyrproject-rtos/zephyr/commit/8ab82f9538b1782fb8bb891a0d33a164dab211bb#r2392594184) this comment is still outstanding
> > 
> > 
> > I'm not sure which comment the link is refering to(broken?) but I did some modifications and replied to one of your previous comments. Do you have an opinion on that? Thanks
> 
> Let me try getting a link again [8ab82f9#diff-e77c05676791af56edbb70e77ae08a9d2a29a04fb33642c40dc716f2198aa555R48](https://github.com/zephyrproject-rtos/zephyr/commit/8ab82f9538b1782fb8bb891a0d33a164dab211bb#diff-e77c05676791af56edbb70e77ae08a9d2a29a04fb33642c40dc716f2198aa555R48) it's the characters, let's not add a hacky incomplete character set and stick to numerical only. Hmm seems to be a github bug, if I open the link manually it goes to the line and shows the comments, if I click the same link but when I've posted on github, it just goes to the commit, without comments, highlights the line but does not scroll to it, basically click commits, click the second commit, scroll down to line 48 in the .c file

Fine. Let me remove the alphabets

heronet committed:
drivers: auxdisplay: Add TM1637 7-segment display driver

Add a new auxdisplay driver for TM1637 7-segment LED displays.

The driver supports:
- 4-digit 7-segment display output
- Decimal point positioning
- Brightness control (0-7 levels)
- Display on/off control
- All digits (0-9)
- Basic cursor positioning

The driver implements the standard AUXDISPLAY API

Signed-off-by: Siratul Islam <sirat4757@gmail.com>

heronet committed:
tests: drivers: auxdisplay: Add TM1637 build test

Add build test coverage for the TM1637 auxdisplay driver.

Creates gpio_devices.overlay for GPIO-based auxdisplay devices
and adds corresponding test case to verify the driver builds
correctly on native_sim platforms.

Signed-off-by: Siratul Islam <sirat4757@gmail.com>

thedjnK commented:
@simonguinot please review

kartben commented:
@JarmouniA please don't dismiss people's reviews. It's for the assignee to do.

simonguinot reviewed (approved):
It looks good to me. Sorry for the delay.
